### PR TITLE
core: refactor handle config update with ConfigSelector

### PR DESCRIPTION
**Pull request by dapengzhang0**
Get rid of `ServiceConfigInterceptor` and handle config update with `ConfigSelector` inside `ClientCallImpl`.

Tests will pass once #7362 is merged.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/ClientCallImpl.java [hunk 1]
```diff
@@ -154,7 +160,8 @@ ClientStream newStream(
         MethodDescriptor<?, ?> method,
         CallOptions callOptions,
         Metadata headers,
-        Context context);
+        Context context,
+        @Nullable MethodInfo methodInfo);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ClientCallImpl.java:164]: As discussed previously, let's keep this interface the same and pass the MethodInfo via CallOptions. (It is fine to continue using the separate keys like RETRY_POLICY_KEY for this PR, but long-term we'll swap to a single call option.)

File: core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java [hunk 2]
```diff
@@ -162,28 +164,37 @@ static ManagedChannelServiceConfig fromServiceConfig(
...
+    if (serviceMap.isEmpty() && serviceMethodMap.isEmpty() && defaultMethodConfig == null) {
+      return null;
+    }
+    return new InternalConfigSelector() {
+      @Override
+      public Result selectConfig(PickSubchannelArgs args) {
+        MethodInfo methodInfo = getMethodConfig(args.getMethodDescriptor());
+        return Result.newBuilder()
+            .setConfig(
+                new ManagedChannelServiceConfig(
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java:187]: Why not use `this`? Was this from an earlier revision and wasn't updated?

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 3]
```diff
@@ -1535,16 +1524,26 @@ public void run() {
...
             // Otherwise, try to use the default config if available
             if (validServiceConfig != null) {
               effectiveServiceConfig = validServiceConfig;
-              configSelector.set(resolvedConfigSelector);
+              if (resolvedConfigSelector != null) {
+                configSelector.set(resolvedConfigSelector);
+                if (effectiveServiceConfig.getDefaultConfigSelector() != null) {
+                  channelLogger.log(
+                      ChannelLogLevel.INFO,
+                      "Method configs in service config will be discarded due to presence of"
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:1538]: I'm not wild about this channel logging, as it seems like spam and is unlikely to help notice a bug. But we can leave it for now and remove it if it becomes a problem.

**Review by ejona86 — APPROVED** [review]

File: core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java (hunk 2 above)
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java:187]: The new instance strips off all configs (e.g. lb config) other than method configs from `this`. The design seems to say the `Result.config` is per RPC config.

File: core/src/main/java/io/grpc/internal/ClientCallImpl.java (hunk 1 above)
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ClientCallImpl.java]: Done.

File: core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java (hunk 2 above)
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java:188]: It seems to say, "this is necessary." If we want to say, "only the per-RPC config will be used" then we can use a comment to do that.

I'm really not wild with calling getMethodConfig() here. We should just return `this` and have this be a trivial implementation.

Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java]: Now I put entire service config in `setConfig()`. Now that the config selector contains all the information of the ManagedConfigServiceConfig, I rename `ManagedConfigServiceConfig.getDefaultConfigSelector()` as `ManagedConfigServiceConfig.asConfigSelector()`.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 3 above)
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: With the new implementation, the converted config selector will never be null. Removed this block.

File: core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java (hunk 2 above)
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java]: You removed the "return null when a config selector is unnecessary" optimization. That means now `configSelector.set(null)` is only used for shutdown when there are no calls and error cases. That's asking for bugs. Restore the return-null optimization or prevent the config selector from ever being null.

Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java]: Good point. Didn't think about it. Restored return-null optimization.

**Review by ejona86 — APPROVED** [review]

**Pull request merged**